### PR TITLE
SWATCH-2101: Retry processing billable_usage's with retry_after field populated

### PR DESCRIPTION
**Pull request by Sgitario**
Jira issue: [SWATCH-2101](https://issues.redhat.com/browse/SWATCH-2101)

## Description
- Query billable_usage_remittance for retry_after > as_of (and non_null) Produce kafka message onto
- platform.billable_usage again Clear retry_after column in the table
- cron job to call the api every 1 hr

## Testing
1. podman-compose up
2. start tally service: `./gradlew :bootRun`
3. insert billable usage remittance:

INSERT INTO billable_usage_remittance(org_id, product_id, metric_id, accumulation_period, sla, usage, billing_provider, billing_account_id, remittance_pending_date, retry_after, remitted_pending_value)
VALUES ('org123', 'rosa', 'CORES', '2024-02', 'Premium', 'Production', 'azure', '123456', '2024-02-11T04:00:09.620406Z', '2023-12-11T04:00:09.620406Z', '4.0')

4. invoke new endpoint:

```
http POST :8000/api/rhsm-subscriptions/v1/internal/rpc/remittance/processRetries \
Origin:console.redhat.com \
x-rh-swatch-psk:placeholder \
x-rh-identity:$(echo -n '{"identity":{"account_number":"","type":"User","user":{"is_org_admin":true},"internal":{"org_id":"16790890"}}}' | base64 -w 0)
```

You should see a new trace in the app logs:

```
2024-02-05 11:59:32,146 [thread=http-nio-8000-exec-1] [INFO ] [org.candlepin.subscriptions.security.LogPrincipalFilter] self- Internal API: /api/rhsm-subscriptions/v1/internal/rpc/remittance/processRetries requested by user: self
2024-02-05 11:59:32,198 [thread=http-nio-8000-exec-1] [INFO ] [org.candlepin.subscriptions.tally.billing.admin.InternalBillingResource] self- Retry billable usage remittances as of 2024-02-05T11:59:32.198623258Z
```

Also, a new message in the following topic: http://localhost:3030/#/cluster/default/topic/n/platform.rhsm-subscriptions.billable-usage/

```
Key
Value
org_id org123
billing_provider azure
billing_account_id 123456
snapshot_date 2024-02-11T04:00:09.620406Z
product_id rosa
sla Premium
usage Production
uom CORES
value 4
```

And finally, the retry_after column for the entry you saved in the step 3, should be NULL:

```
select retry_after from billable_usage_remittance where org_id='org123';
```

### CODE REVIEW

File: src/main/resources/liquibase/202402051200-add-id-and-hardware-columns-to-billable_usage_remittance-table.xml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  <changeSet id="202401151414-01" author="lburnett">
+
+    <comment>Add billable_usage_remittance.tally_id and billable_usage_remittance.hardware_measurement_type columns.</comment>
+    <addColumn tableName="billable_usage_remittance">
+      <column name="tally_id" type="UUID"/>
+      <column name="hardware_measurement_type" type="VARCHAR(32)"/>
+    </addColumn>
+    <rollback>
+      <dropColumn tableName="billable_usage_remittance" columnName="tally_id"/>
+      <dropColumn tableName="billable_usage_remittance" columnName="hardware_measurement_type"/>
```
Comment (Sgitario) [inline src/main/resources/liquibase/202402051200-add-id-and-hardware-columns-to-billable_usage_remittance-table.xml:16]: @kflahert I had to add these two columns to be able to replicate the billing usage message into the topic.

File: src/main/java/org/candlepin/subscriptions/tally/billing/admin/InternalBillingController.java [hunk 2]
```diff
@@ -62,6 +69,37 @@ public List<MonthlyRemittance> process(BillableUsageRemittanceFilter filter) {
...
+        .withId(remittance.getTallyId())
+        .withSnapshotDate(remittance.getKey().getRemittancePendingDate())
+        .withProductId(remittance.getKey().getProductId())
+        .withSla(BillableUsage.Sla.fromValue(remittance.getKey().getSla()))
+        .withUsage(BillableUsage.Usage.fromValue(remittance.getKey().getUsage()))
+        .withBillingProvider(
+            BillableUsage.BillingProvider.fromValue(remittance.getKey().getBillingProvider()))
+        .withBillingAccountId(remittance.getKey().getBillingAccountId())
+        .withUom(remittance.getKey().getMetricId())
+        .withValue(remittance.getRemittedPendingValue())
```
Comment (Sgitario) [inline src/main/java/org/candlepin/subscriptions/tally/billing/admin/InternalBillingController.java:99]: @kflahert is this the correct value to use?

File: src/main/java/org/candlepin/subscriptions/tally/billing/admin/InternalBillingController.java [hunk 3]
```diff
@@ -62,6 +69,37 @@ public List<MonthlyRemittance> process(BillableUsageRemittanceFilter filter) {
...
+    // to save the retry after column for all the entities
+    remittanceRepository.saveAll(remittances);
+    return remittances.size();
+  }
+
+  private BillableUsage toBillableUsage(BillableUsageRemittanceEntity remittance) {
+    return new BillableUsage()
+        .withOrgId(remittance.getKey().getOrgId())
+        .withId(remittance.getTallyId())
+        .withSnapshotDate(remittance.getKey().getRemittancePendingDate())
```
[the remark this answers is not in the document]
Comment (kflahert) [reply src/main/java/org/candlepin/subscriptions/tally/billing/admin/InternalBillingController.java:91]: Yeah, remittancePendingDate is good.

File: src/main/java/org/candlepin/subscriptions/tally/billing/admin/InternalBillingController.java (hunk 2 above)
Comment (kflahert) [reply src/main/java/org/candlepin/subscriptions/tally/billing/admin/InternalBillingController.java:99]: Yep, this looks good.

**Review by TrayvonMcKnight — APPROVED** [review]
https://drive.google.com/drive/folders/1kv0s8ZgrQW6yf6cK5rHtGKB8ZBbsE3OP?usp=sharing

**Pull request merged**